runtime/compiler/runtime/Listener.cpp

```diff
@@ -20,18 +20,284 @@
+
+   bio = (*OBIO_new_ssl)(sslCtx, false);
+   if (!bio)
+      return handleOpenSSLConnectionError(connfd, ssl, bio, "Error creating new BIO");
+
+   if ((*OBIO_ctrl)(bio, BIO_C_SET_SSL, true, (char *)ssl) != 1) // BIO_set_ssl(bio, ssl, true)
+      return handleOpenSSLConnectionError(connfd, ssl, bio, "Error setting BIO SSL");
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "SSL connection on socket 0x%x, Version: %s, Cipher: %s\n",
+                                                     connfd, (*OSSL_get_version)(ssl), (*OSSL_get_cipher)(ssl));
+   return true;
+   }
 
 TR_Listener::TR_Listener()
    : _listenerThread(NULL), _listenerMonitor(NULL), _listenerOSThread(NULL), 
    _listenerThreadAttachAttempted(false), _listenerThreadExitFlag(false)
    {
    }
 
+void
+TR_Listener::serveRemoteCompilationRequests(BaseCompileDispatcher *compiler)
+   {
+   TR::PersistentInfo *info = getCompilationInfo(jitConfig)->getPersistentInfo();
+   SSL_CTX *sslCtx = NULL;
+   if (JITServer::CommunicationStream::useSSL())
+      {
+      JITServer::CommunicationStream::initSSL();
+      sslCtx = createSSLContext(info);
+      }
+
+   uint32_t port = info->getJITServerPort();
+   uint32_t timeoutMs = info->getSocketTimeout();
+   struct pollfd pfd = {0};
+   int sockfd = socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+   if (sockfd < 0)
+      {
+      perror("can't open server socket");
+      exit(1);
+      }
+
+   // see `man 7 socket` for option explanations
+   int flag = true;
+   if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, (void *)&flag, sizeof(flag)) < 0)
+      {
+      perror("Can't set SO_REUSEADDR");
+      exit(-1);
+      }
+   if (setsockopt(sockfd, SOL_SOCKET, SO_KEEPALIVE, (void *)&flag, sizeof(flag)) < 0)
+      {
+      perror("Can't set SO_KEEPALIVE");
+      exit(-1);
+      }
+
+   struct sockaddr_in serv_addr;
+   memset((char *)&serv_addr, 0, sizeof(serv_addr));
+   serv_addr.sin_family = AF_INET;
+   serv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+   serv_addr.sin_port = htons(port);
+
+   if (bind(sockfd, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0)
+      {
+      perror("can't bind server address");
+      exit(1);
+      }
+   if (listen(sockfd, SOMAXCONN) < 0)
+      {
+      perror("listen failed");
+      exit(1);
+      }
+
+   pfd.fd = sockfd;
+   pfd.events = POLLIN;
+
+   while (!getListenerThreadExitFlag())
+      {
+      int32_t rc = 0;
+      struct sockaddr_in cli_addr;
+      socklen_t clilen = sizeof(cli_addr);
+
+      rc = poll(&pfd, 1, OPENJ9_LISTENER_POLL_TIMEOUT);
+      if ((rc == 0) && getListenerThreadExitFlag())
+         {
+         break;
+         }
+      else if (rc < 0)
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Error in polling socket: errno=%d", errno);
+	 exit(1);
```

Comment:
For a hard exit like this I would prefer a message to stderr, rather than the vlog (or in addition to the vlog). We used `perror` below, so we could use here as well.

Reply:
Done.